File: content/csharp/learn/01-classes-and-properties/solution/Account.cs
namespace Lesson;

public record Money(decimal Amount, string Currency);

public class Account
{
    public string Owner { get; }

    // Readable everywhere, writable only in here: the invariants below are
    // worth nothing if a caller can assign the balance directly.
    public decimal Balance { get; private set; }

    public Account(string owner, decimal opening = 0)
    {
        if (string.IsNullOrWhiteSpace(owner))
        {
            throw new ArgumentException("owner must not be blank", nameof(owner));
        }
        Owner = owner;
        Balance = opening;
    }

    public void Deposit(decimal amount)
    {
        if (amount <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(amount), "deposit must be positive");
        }
        Balance += amount;
    }

    public void Withdraw(decimal amount)
    {
        if (amount <= 0)
        {
            throw new ArgumentOutOfRangeException(nameof(amount), "withdrawal must be positive");
        }
        if (amount > Balance)
        {
            throw new InvalidOperationException("insufficient funds");
        }
        Balance -= amount;
    }

    public Money AsMoney() => new Money(Balance, "EUR");
}

File: content/java/silver/05-methods-and-objects/solution/Account.java
public class Account {

    private final String owner;
    private double balance;

    public Account(String owner, double balance) {
        this.owner = owner;
        this.balance = balance;
    }

    public Account(String owner) {
        this(owner, 0);
    }

    public String getOwner() {
        return owner;
    }

    public double getBalance() {
        return balance;
    }

    public void deposit(double amount) {
        if (amount <= 0) {
            throw new IllegalArgumentException("deposit must be positive");
        }
        balance += amount;
    }

    public void withdraw(double amount) {
        if (amount <= 0) {
            throw new IllegalArgumentException("withdrawal must be positive");
        }
        if (amount > balance) {
            throw new IllegalArgumentException("insufficient funds");
        }
        balance -= amount;
    }

    public void transferTo(Account other, double amount) {
        withdraw(amount);
        other.deposit(amount);
    }

    public static void main(String[] args) {
        Account a = new Account("ada", 100);
        Account b = new Account("bob");
        a.transferTo(b, 25);
        System.out.println(a.getBalance() + " " + b.getBalance());
    }
}
